### PR TITLE
Unify caching and fix adjoint tests

### PR DESCRIPTION
**Pull request by gautierronan**
This PR unifies the way caching is handled by grouping everything inside a generalized `TDTensor` class. This also fixes the adjoint tests that were broken, see https://github.com/pierreguilmin/dynamiqs/issues/128.

This is not super well polished yet, I mostly wanted to see if you guys agreed with the way this is written. If you do I will add back the tests for caching. If not, I'm completely open to other proposals.

### CODE REVIEW

**Comment by gautierronan** [comment]
Here's a second version because I was not happy with the initial one. In the end I used a simple `lru_cache` decorator like @abocquet was proposing initially for the Kraus operators, and for functions that only depend on time, I use the generalized `TDTensor` class. For the issue with caching grad or no_grad tensors, I defined a `clear_cache` function which is called once at the end of every `run_nograd`.

File: dynamiqs/solvers/solver.py [hunk 1]
```diff
@@ -128,6 +101,16 @@ def _save_exp_ops(self, y: Tensor):
         if len(self.exp_ops) > 0:
             self.exp_save[..., self.save_counter] = bexpect(self.exp_ops, y)
 
+    def clear_cache(self):
```
Comment (pierreguilmin) [inline dynamiqs/solvers/solver.py:104]: This is quite a dangerous way to do what you want, let's keep this for now, I'll propose something cleaner soon.

File: dynamiqs/solvers/solver.py [hunk 2]
```diff
@@ -128,6 +101,16 @@ def _save_exp_ops(self, y: Tensor):
...
+        """Clear the cache of all `TDTensor` attributes."""
+        for attr_str in dir(self):
+            attr = getattr(self, attr_str)
+            if isinstance(attr, TDTensor):
+                attr.clear_cache()
+
+            if hasattr(attr, 'cache_clear'):
+                attr.cache_clear()
```
Comment (pierreguilmin) [inline dynamiqs/solvers/solver.py:112]: Why adding this?

File: dynamiqs/mesolve/rouchon.py [hunk 3]
```diff
@@ -20,26 +21,69 @@ def __init__(self, *args, **kwargs):
...
         # -> (b_H, 1, n, n)
         return self.I - 1j * dt * self.H_nh(t)
 
-    @depends_on_H
-    def M0_adj(self, t: float, dt: float) -> Tensor:
+    @lru_cache(maxsize=1)
+    def M1s(self, t: float, dt: float) -> Tensor:
         # -> (b_H, 1, n, n)
-        return self.I + 1j * dt * self.Hdag_nh(t)
+        return sqrt(dt) * self.jump_ops
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply dynamiqs/mesolve/rouchon.py:86]: Actually the Rouchon solver is a fixed time step scheme so `dt` is always the same, we don't need to cache on dt anywhere in this file.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
I like the idea!!

**Comment by pierreguilmin** [comment]
Replaced by #133.

**Pull request closed without merging**